Detectors/ITSMFT/common/reconstruction/src/Clusterer.cxx

```diff
@@ -236,81 +236,101 @@ void Clusterer::ClustererThread::finishChip(ChipPixelData* curChipData, CompClus
-        }
-      } else {
-        if ((rowSpanW = ClusterPattern::MaxPatternBits / colSpan) == 0) {
-          rowSpanW = 1;
-          colSpanW = ClusterPattern::MaxPatternBits;
-        }
-      }
+
+    auto chipID = curChipData->getChipID();
+    uint16_t colSpan = (colMax - colMin + 1);
+    uint16_t rowSpan = (rowMax - rowMin + 1);
+    if (colSpan <= o2::itsmft::ClusterPattern::MaxColSpan &&
+        rowSpan <= o2::itsmft::ClusterPattern::MaxRowSpan) {
+      streamCluster(rowMin, rowSpan, colMin, colSpan, npix, chipID,
+                    compClusPtr, patternsPtr, labelsClusPtr, nlab);
+    } else {
+      LOG(WARNING) << "Splitting a huge cluster !  ChipID: " << chipID;
+
+      colSpan %= o2::itsmft::ClusterPattern::MaxColSpan;
+      if (colSpan == 0)
+        colSpan = o2::itsmft::ClusterPattern::MaxColSpan;
+
+      rowSpan %= o2::itsmft::ClusterPattern::MaxRowSpan;
+      if (rowSpan == 0)
+        rowSpan = o2::itsmft::ClusterPattern::MaxRowSpan;
+
+      do {
+        uint16_t r = rowMin, rsp = rowSpan;
+
+        do {
+          streamCluster(r, rsp, colMin, colSpan, npix, chipID,
+                        compClusPtr, patternsPtr, labelsClusPtr, nlab);
+          r += rsp;
+          rsp = o2::itsmft::ClusterPattern::MaxRowSpan;
+        } while (r < rowMax);
+
+        colMin += colSpan;
+        colSpan = o2::itsmft::ClusterPattern::MaxColSpan;
+      } while (colMin < colMax);
     }
+  }
+}
 
-    if (labelsClusPTr) { // MC labels were requested
-      auto cnt = compClusPtr->size();
-      for (int i = nlab; i--;) {
-        labelsClusPTr->addElement(cnt, labelsBuff[i]);
-      }
+void Clusterer::ClustererThread::streamCluster(uint16_t rowMin, uint16_t rowSpanW, uint16_t colMin, uint16_t colSpanW, int npix, uint16_t chipID, CompClusCont* compClusPtr, PatternCont* patternsPtr, MCTruth* labelsClusPtr, int nlab)
+{
+  if (labelsClusPtr) { // MC labels were requested
+    auto cnt = compClusPtr->size();
+    for (int i = nlab; i--;) {
+      labelsClusPtr->addElement(cnt, labelsBuff[i]);
     }
+  }
 
-    // add to compact clusters, which must be always filled
-    unsigned char patt[ClusterPattern::MaxPatternBytes] = {0}; // RSTODO FIX pattern filling
-    for (int i = 0; i < npix; i++) {
-      const auto pix = pixArrBuff[i];
-      unsigned short ir = pix.getRowDirect() - rowMin, ic = pix.getCol() - colMin;
-      if (ir < rowSpanW && ic < colSpanW) {
-        int nbits = ir * colSpanW + ic;
-        patt[nbits >> 3] |= (0x1 << (7 - (nbits % 8)));
-      }
+  // add to compact clusters, which must be always filled
+  unsigned char patt[ClusterPattern::MaxPatternBytes] = {0}; // RSTODO FIX pattern filling
+  for (int i = 0; i < npix; i++) {
+    const auto pix = pixArrBuff[i];
+    unsigned short ir = pix.getRowDirect() - rowMin, ic = pix.getCol() - colMin;
+    if (ir < rowSpanW && ic < colSpanW) {
+      int nbits = ir * colSpanW + ic;
+      patt[nbits >> 3] |= (0x1 << (7 - (nbits % 8)));
```

Comment:
If I understand correctly, for large clusters you want to loop over the same npix pixels pixArrBuff, each time selecting only the pixels belonging to the updated bounding box? But then for some pixels row / col could be < updated rowMin / colMin, so, before calculating ir, ic, a check will be needed. 
Apart from that, if the initial cluster BB was, say, 129 by 129, your splitting would 1st consider the BB of 1 by 1 in the upper left corner of the original BB. It is not guaranteed that there will be a fired pixel there, it may happen that no cluster will be added with some of the "partial BBs", I think it should be checked after this loop before proceding to cluster pattID calculation/storage. For the same reason, the MC labels would need to be added only after making sure that there is a sub-cluster to add.
Given that this method of large clusters will require more checks, perhaps it is worth to make a dedicated version for large clusters, to not create an overhead for the bulk?

Reply:
@shahor02  Indeed, a few extra checks are needed.  Thank you for catching it !
I will then provide a separate (slower) function for dealing with pieces of huge clusters.